### PR TITLE
[KHRGA-85] Updated EP nd item class

### PR DESCRIPTION
**Pull request by imvucic**
Issue: KHRGA-85

### CODE REVIEW

File: source/iface/nd_item.rst [hunk 1]
```diff
@@ -12,174 +10,259 @@
...
    class nd_item;
 
-The ``nd_item`` describes the location of a point in an
-:ref:`nd_range`.
-
-An ``nd_item`` is typically passed to a kernel function in a
-:ref:`handler-parallel_for`. It addition to containing the ``id`` of
-the work item in the work group and global space, the ``nd_item`` also
-contains the :ref:`nd_range` defining the index space.
+The ``nd_item`` identifies an instance of the function object
```
Comment (mkinsner) [inline source/iface/nd_item.rst:16]:
```suggestion
The ``nd_item`` class identifies an instance of the function object
```

File: source/iface/nd_item.rst [hunk 2]
```diff
@@ -12,174 +10,259 @@
...
+position within the current work-group.
+
+::
 
-Returns id for the point in the work group.
+  size_t get_local_id(int dimension) const
+
+Return the constituent element of the local id representing the
+work-item’s position within the current
+work-group in the given ``Dimension``.
```
Comment (mkinsner) [inline source/iface/nd_item.rst:77]:
```suggestion
work-group in the given ``dimension``.
```

File: source/iface/nd_item.rst [hunk 3]
```diff
@@ -12,174 +10,259 @@
...
+Return the constituent work-group, group representing the work-group's
+position within the overall ``nd-range``.
+
+::
+
+  size_t get_group(int dimension) const
+
+Return the constituent element of the group ``id`` representing
+the work-group’s position within the overall
+``nd_range`` in the given ``Dimension``.
```
Comment (mkinsner) [inline source/iface/nd_item.rst:107]:
```suggestion
``nd_range`` in the given ``dimension``.
```

File: source/iface/nd_item.rst [hunk 4]
```diff
@@ -12,174 +10,259 @@
...
+  range<Dimensions> get_group_range() const
+
+Returns the number of work-groups in the iteration space.
+
+::
+
+  size_t get_group_range(int dimension) const
 
-Returns the number of groups in every dimension.
+Return the number of work-groups for ``Dimension`` in
```
Comment (mkinsner) [inline source/iface/nd_item.rst:142]:
```suggestion
Return the number of work-groups for ``dimension`` in
```

File: source/iface/nd_item.rst [hunk 5]
```diff
@@ -12,174 +10,259 @@
...
+                                     size_t numElements, size_t destStride) const
 
-  template <sycl::access::mode accessMode = sycl::access::mode::read_write>
-  void mem_fence(sycl::access::fence_space accessSpace =
-    sycl::access::fence_space::global_and_local) const;
+All four ``async_work_group_copy`` functions listed above are deprecated
+in SYCL 2020. Has the same effect as the overload taking
+``decorated_local_ptr`` and ``decorated_global_ptr``
+except that the ``dest`` and ``src`` parameters are
+``multi_ptr#s with [code]#access::decorated::legacy``.
```
Comment (mkinsner) [inline source/iface/nd_item.rst:230]: Formatting is broken here

File: source/iface/nd_item.rst (hunk 1 above)
Comment (imvucic) [reply source/iface/nd_item.rst]: Fixed with commit https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/96f7087e77284d0cd88da5b0ac8d3c8f3888d50e

File: source/iface/nd_item.rst [hunk 6]
```diff
@@ -12,174 +10,259 @@
...
+position in the global iteration space.
 
-Returns global :ref:`id` for the requested dimensions.
+
+::
+
+  size_t get_global_id(int dimension) const
+
+Return the constituent element of the global id representing
+the work-item’s position in the ``nd-range`` in the given ``Dimension``.
```
[the remark this answers is not in the document]
Comment (imvucic) [reply source/iface/nd_item.rst]: Fixed with commit https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/96f7087e77284d0cd88da5b0ac8d3c8f3888d50e

File: source/iface/nd_item.rst (hunk 2 above)
Comment (imvucic) [reply source/iface/nd_item.rst]: Fixed with commit https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/96f7087e77284d0cd88da5b0ac8d3c8f3888d50e

File: source/iface/nd_item.rst (hunk 3 above)
Comment (imvucic) [reply source/iface/nd_item.rst]: Fixed with commit https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/96f7087e77284d0cd88da5b0ac8d3c8f3888d50e

File: source/iface/nd_item.rst (hunk 4 above)
Comment (imvucic) [reply source/iface/nd_item.rst]: Fixed with commit https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/96f7087e77284d0cd88da5b0ac8d3c8f3888d50e

File: source/iface/nd_item.rst (hunk 5 above)
Comment (imvucic) [reply source/iface/nd_item.rst:230]: Could you outline more details? I see in output it looks fine but didn't catch wrong part...

Comment (mkinsner) [reply source/iface/nd_item.rst:230]: image

In my render, there are '#' and '[code]' in the output, and the word "with" is in code font when I don't think it should be.  Are you seeing differently in your render?

Comment (imvucic) [reply source/iface/nd_item.rst:230]: @mkinsner could you recheck is spec correct?
[https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.members.nditem](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.members.nditem)

Deprecated in SYCL 2020. Has the same effect as the overload taking decorated_local_ptr and decorated_global_ptr except that the dest and src parameters are multi_ptr#s with [code]#access::decorated::legacy.


Please put here corrected version, I was thinking that spec is fine...
The same problem for group class in [
Table 118. Member functions for the group class
](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.members.group)

Comment (mkinsner) [reply source/iface/nd_item.rst:230]: Thanks for pointing this out.  I've opened a PR to fix the spec:
https://github.com/KhronosGroup/SYCL-Docs/pull/518

Comment (imvucic) [reply source/iface/nd_item.rst]: Removed 's' with : https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/6bdee50f6363fd9a9e1fb925605895e377ce0b13

Comment (imvucic) [reply source/iface/nd_item.rst]: Done with https://github.com/KhronosGroup/SYCL_Reference/pull/122/commits/fcc33dcf4832c0c5cbf1d220d17bafb05b972616

**Review by mkinsner — APPROVED** [review]

**Pull request merged**